rfd/0039-session-encryption.md

```diff
@@ -0,0 +1,105 @@
+
+Broadly, there are two options available:
+
+1. Sharing a single encryption key across all authentication servers.
+2. Taking a card from the AWS Encryption SDK playbook and using a two layer key system.
+
+Option 1 has multiple significant drawbacks which make it unsuitable for production use. The major issue is that all data is fundamentally tied to the only layer of security meaning that if a malicous party potentially acquired access hold to the key, all data would have to be reincrypted. This makes key rotation prohibitive. Another major issue is that of prolonged key reuse which contradicts industry best practices.
+
+Option 2 is doing something very similar to what the AWS Encryption SDK does. Ideally using that library would be ideal since it is a very nice encryption toolkit but unfortunately it isn't available for Go. The strategy involves using unique ephemeral data keys for each encrypted object (rotated when reincrypted/modified) that are encrypted and bundled with each encrypted object in a defined format. A central PEM encoded master keypair stored in a new resource is used to encrypt these per object keys.
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.key` stored next to the encrypted object.
+
+Each encrypted object has an accompanying file named `$objectName.key` that stores a JSON document of this format.
+
+```json
+{
+    // Contains one or more base64 encoded AES-CBC-HMACSHA256 data key encrypted with the master key
+    // together with a unique identifier for the related master key.
+    "dataKey": "[]{
+        key: string,
+        masterKey: string
+    }",
+
+    // An integer starting at 1 that communicates the schema used for encryption.
+    // May be changed in the future.
+    "version": "number",
+}
+```
+
+This format allows a decrypting auth server to easily fetch and decrypt the data key using the master key. It can then fetch and decrypt the object itself.
+Key rotation in event of a security incident can be performed by fetching and rewriting every key file to reencrypt the data key.
+
+The master key resource stores multiple copies of the master key in the following form where the map key is a unique per-auth identifier and the value is a base64 encoded encrypted master key. The master key is encrypted with a asymmetric keypair that never leaves the auth server. This integrates well with HSM and provides good security.
```

Comment:
> The master key is encrypted with a asymmetric keypair that never leaves the auth server

I assume in the non-HSM case these are essentially stored on disk, and that if all auth servers "lose" their keypairs, all encrypted data will be lost? I wonder if it is necessary to encrypt the master key before writing it to the backend, I know we store other private keys in the backend unencrypted

Reply:
Hmm, I think support for this was requested. If we don't do this, HSM's serve no real purpose. They could be used for encrypting and decrypting with the master key but not all HSM's support that operation. Would just storing the master key in the backend unencrypted still meet requirements? @russjones
The issue isn't super clear on them.